### ISSUE TITLE
PostgreSQL plugin loses data on pressure

### ISSUE THREAD

**Issue by Funzinator** [fluent/fluent-bit#2112]
## Bug Report

**Describe the bug**

I have prepared a big file with lines of JSON data. If I use pgout, only ~8 % of the data gets stored.
If I process it with stdout output, the amount of lines matches up.

**To Reproduce**
- Rubular link if applicable: n/a
- Example log message if applicable:
```
Fluent Bit v1.4.2
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/04/18 00:06:02] [ info] [storage] version=1.0.3, initializing...
[2020/04/18 00:06:02] [ info] [storage] in-memory
[2020/04/18 00:06:02] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/04/18 00:06:02] [ info] [engine] started (pid=10133)
[2020/04/18 00:06:02] [ info] [out_pgsql] host=xxx port=5432 dbname=fluent_test ...
[2020/04/18 00:06:02] [ info] [out_pgsql] host=xxx port=5432 dbname=fluent_test OK
[2020/04/18 00:06:02] [ info] [out_pgsql] we check that the table "fluent_test" exists, if not we create it
[2020/04/18 00:06:02] [ info] [out_pgsql] switching postgresql connection to non-blocking mode
[2020/04/18 00:06:02] [ info] [sp] stream processor started
[2020/04/18 00:06:09] [ warn] [out_pgsql] another command is already in progress
[2020/04/18 00:06:10] [ warn] [out_pgsql] another command is already in progress
[2020/04/18 00:06:10] [ warn] [out_pgsql] another command is already in progress
[2020/04/18 00:06:13] [ warn] [out_pgsql] another command is already in progress
[...]
[2020/04/18 00:07:10] [ warn] [input:stdin:stdin.0] end of file (stdin closed by remote end)
[2020/04/18 00:07:11] [ warn] [out_pgsql] another command is already in progress
[2020/04/18 00:07:11] [ warn] [engine] service will stop in 5 seconds
[2020/04/18 00:07:16] [ info] [engine] service stopped
```
- Steps to reproduce the problem:

create `test.sh`:
```
#!/bin/bash

for i in {1..1000000} ; do
  echo "{\"key\": \"some value\",\"key\": \"some value XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXxxXXXx\"}"
done
```
then execute like this:
```
bash tesh.sh > test.json
cat test.json | fluent-bit -i stdin -o pgsql -p Host=xxx -p Port=5432 -p Database=fluent_test -p Table=fluent_test -p User=fluent_test -p Password=xxx -s 24576 -v  
```
After that, `select count(*) from fluent_test` shows 177867 lines but in fact 1000000 lines of JSON were fed to pgout.

**Expected behavior**
All records to be stored in PostgreSQL table

**Your Environment**
* Version used: 1.4.2
* Configuration: n/a
* Server type and version: FreeBSD
* Operating System and version: 11.3
* Filters and plugins: no filters, pgout plugin

**Comment by Funzinator** [comment]
In case it helps, I have added a counter property to my JSON data and sorted then the data that ended up in the database:
`select data->>'log_counter' from fluent_test order by data->>'log_counter' asc;`
gets the following:
```
0
1
10
100
1000
1001
1002
1003
1003648
1003649
1003650
...
```
So the first two are persisted, then 8 are dropped and it goes on like this

**Comment by sxd** [comment]
@Funzinator thank for reporting this! We will try to reproduce and fix it!

**Comment by danboucher78** [comment]
Hi
I experiencing the same issue.  Here is the way to reproduce the problem:
`#!/bin/sh
while :; do
    echo -n "{\"key\": \"some value\"}"
    sleep 1
done`

**Comment by danboucher78** [comment]
Hi

I'm experiencing the same issue.  Here is the way I reproduce it:

Create test.sh and make it executable
```
#!/bin/sh

while :; do
	echo -n "{\"key\": \"some value\"}"
	sleep 1
done
```
then run like this
```
./test.sh | fluent-bit -f 1 -i stdin -o pgsql -p Host=x.x.x.x -p Port=5432 -p Database=fluentbit -p Table=fluentbittest -p User=fluentbit -p Password=fluentbit -v
```
Here is the log from Fluentbit
```
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/04/24 11:12:33] [ info] Configuration:
[2020/04/24 11:12:33] [ info]  flush time     | 1.000000 seconds
[2020/04/24 11:12:33] [ info]  grace          | 5 seconds
[2020/04/24 11:12:33] [ info]  daemon         | 0
[2020/04/24 11:12:33] [ info] ___________
[2020/04/24 11:12:33] [ info]  inputs:
[2020/04/24 11:12:33] [ info]      stdin
[2020/04/24 11:12:33] [ info] ___________
[2020/04/24 11:12:33] [ info]  filters:
[2020/04/24 11:12:33] [ info] ___________
[2020/04/24 11:12:33] [ info]  outputs:
[2020/04/24 11:12:33] [ info]      pgsql.0
[2020/04/24 11:12:33] [ info] ___________
[2020/04/24 11:12:33] [ info]  collectors:
[2020/04/24 11:12:33] [debug] [storage] [cio stream] new stream registered: stdin.0
[2020/04/24 11:12:33] [ info] [storage] version=1.0.3, initializing...
[2020/04/24 11:12:33] [ info] [storage] in-memory
[2020/04/24 11:12:33] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/04/24 11:12:33] [ info] [engine] started (pid=1329125)
[2020/04/24 11:12:33] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/04/24 11:12:33] [ info] [out_pgsql] host=10.10.28.125 port=5432 dbname=fluentbit ...
[2020/04/24 11:12:33] [ info] [out_pgsql] host=10.10.28.125 port=5432 dbname=fluentbit OK
[2020/04/24 11:12:33] [ info] [out_pgsql] we check that the table "fluentbittest" exists, if not we create it
NOTICE:  relation "fluentbittest" already exists, skipping
[2020/04/24 11:12:33] [ info] [out_pgsql] switching postgresql connection to non-blocking mode
[2020/04/24 11:12:33] [debug] [router] default match rule stdin.0:pgsql.0
[2020/04/24 11:12:33] [ info] [sp] stream processor started
[2020/04/24 11:12:33] [debug] [task] created task=0x7f6e0b031600 id=0 OK
[2020/04/24 11:12:33] [debug] [task] destroy task=0x7f6e0b031600 (task_id=0)
[2020/04/24 11:12:34] [debug] [task] created task=0x7f6e0b031600 id=0 OK
[2020/04/24 11:12:34] [debug] [task] destroy task=0x7f6e0b031600 (task_id=0)
[2020/04/24 11:12:35] [debug] [task] created task=0x7f6e0b031600 id=0 OK
[2020/04/24 11:12:35] [ warn] [out_pgsql] another command is already in progress

[2020/04/24 11:12:35] [debug] [task] destroy task=0x7f6e0b031600 (task_id=0)
[2020/04/24 11:12:36] [debug] [task] created task=0x7f6e0b031600 id=0 OK
[2020/04/24 11:12:36] [debug] [task] destroy task=0x7f6e0b031600 (task_id=0)
[2020/04/24 11:12:37] [debug] [task] created task=0x7f6e0b031600 id=0 OK
[2020/04/24 11:12:37] [debug] [task] destroy task=0x7f6e0b031600 (task_id=0)
[2020/04/24 11:12:38] [debug] [task] created task=0x7f6e0b031600 id=0 OK
[2020/04/24 11:12:38] [ warn] [out_pgsql] another command is already in progress

[2020/04/24 11:12:38] [debug] [task] destroy task=0x7f6e0b031600 (task_id=0)
```
It does the same with version 1.4.2

**Comment by sxd** [comment]
Hi @Funzinator @danboucher78  and @ticpu 

Yes after a few testing and searching I was able to replicate this issue thanks for the report!

This it's what we found, the main problem comes from the fact that since the connection it's async there's a point that we will try to send data to the PostgreSQL database and the connection will be busy, and by default, it will only try again once.

The proposed fix for this, and that will add some speed up, it's to create a pool of connection, with a default of 5 connection.

I already have a branch working with this that I'll publish later today so everyone here can review it and test, that will be great! all help it's really useful!

I'll review also your patch @ticpu since maybe we can use both method or implement them, everything helps.

**Comment by ticpu** [comment]
With a connection pool you may end up with the same issue is there is contention, I don't see much incentive in handling asynchronous data insertion if fluent is already asynchronous.  The only thing asynchronous might handle better is reconnection is something blocks for too long.

With a connection pool you will also send so many connections to Postgresql that if you've got 100 host you'll have to setup Postgres to handle 500 connections for little benefits and with enough data coming in, you'll hit the same problem.

**Comment by Funzinator** [comment]
I agree with @ticpu on that. PostgreSQL is still quite fast for these insertions into jsonb columns - just not as fast as reading from stdin, of course. The amount of memory to be hold in case of contention is probably not very high and therefore a small price to not losing log entries.

**Comment by sxd** [comment]
Fix with pull request #2236
